runtime/compare.c

```diff
@@ -276,33 +304,57 @@ static intnat do_compare_val(struct compare_stack* stk,
       /* Compare sizes first for speed */
       if (sz1 != sz2) return sz1 - sz2;
       if (sz1 == 0) break;
-      /* Remember that we still have to compare fields 1 ... sz - 1 */
-      if (sz1 > 1) {
-        sp++;
-        if (sp >= stk->limit) sp = compare_resize_stack(stk, sp);
-        sp->v1 = &Field(v1, 1);
-        sp->v2 = &Field(v2, 1);
-        sp->count = sz1 - 1;
+      /* Remember that we still have to compare fields 1 ... sz - 1. */
+      if (sz1 > 1 || CAMLunlikely(signal_poll_timer == 0)) {
+        /* In the special case where we must poll, we always push the
+           whole blocks to the compare stack, and break (the switch)
+           to the poll point below instead of continuing at the
+           top. */
+        if (next_stack_slot >= stk->limit)
+          next_stack_slot = compare_resize_stack(stk, next_stack_slot);
+        struct compare_item* sp = next_stack_slot;
+        next_stack_slot ++;
+        sp->v1 = v1;
+        sp->v2 = v2;
+        sp->size = Val_long(sz1);
+        if (CAMLunlikely(signal_poll_timer == 0)) {
+          sp->offset = Val_long(0);
+          break;
+        } else {
+          sp->offset = Val_long(1);
+        }
```

Comment:
It seems you worked hard to combine two cases of pushing to the stack in one piece of code.  One case is when we need to descend into a pair of blocks, and the other is when we need to poll.  My instinct is to keep both cases distinct.  Before I suggest another code layout, can you tell me why you chose to combine these two cases of pushing?

Reply:
I assumed that code size could be performance-sensitive for this function, and that sharing most of the code of the two cases was better. I am happy to separate the two cases as you suggest, but the obvious way I can think of would duplicate the `push` logic.

Reply:
OK, I'll keep that in mind and think about it.